code-generate/parser.go

```diff
@@ -118,11 +118,120 @@ func createRamlTypeAttribute(name string, properties interface{}) RamlTypeAttrib
+type ResourceService struct {
+	BasePath          string
+	HasCreate         bool
+	HasList           bool
+	ResourceMethods   []ResourceMethod
+	ResourceType      string
+	ResourceQueryType string
+	ResourceDraftType string
+}
+
+func parseYaml(data yaml.MapSlice) (objects []RamlType, resources []ResourceService) {
 	types := getPropertyValue(data, "types")
 	for _, mapItem := range types.(yaml.MapSlice) {
 		obj := createRamlType(mapItem.Key.(string), mapItem.Value.(yaml.MapSlice))
 		objects = append(objects, obj)
 	}
+
+	resources = make([]ResourceService, 0)
+	apiResources := getPropertyValue(data, "/{projectKey}")
+	for _, apiResource := range apiResources.(yaml.MapSlice) {
+		if !strings.HasPrefix(apiResource.Key.(string), "/") {
+			continue
+		}
+		resourceService := ResourceService{
+			BasePath: apiResource.Key.(string)[1:],
+		}
+		for _, item := range apiResource.Value.(yaml.MapSlice) {
+			key := item.Key.(string)
+			if key == "get" {
+				resourceService.HasList = true
+			}
+			if key == "post" {
+				resourceService.HasCreate = true
+			}
+			if key == "type" {
+				typeInfo := getPropertyValue(apiResource.Value.(yaml.MapSlice), "type")
+				if _, ok := typeInfo.(yaml.MapSlice); !ok {
+					continue
+				}
+				baseDomain := getPropertyValue(typeInfo.(yaml.MapSlice), "baseDomain").(yaml.MapSlice)
+				resourceService.ResourceType = getPropertyString(baseDomain, "resourceType")
+				resourceService.ResourceQueryType = strings.Split(getPropertyString(baseDomain, "resourceQueryType"), " ")[0]
+				resourceService.ResourceDraftType = getPropertyString(baseDomain, "resourceDraft")
+			}
+			if strings.HasPrefix(key, "/") {
+				resourceMethod := ResourceMethod{
+					Path: key,
+				}
+				switch key {
+				case "/{ID}":
+					resourceMethod.PathParameterName = "ID"
+				case "/key={key}":
+					resourceMethod.PathParameterName = "key"
+				}
+				for _, methodEntry := range item.Value.(yaml.MapSlice) {
+					methodKey := methodEntry.Key.(string)
+					if methodKey == "(methodName)" {
+						resourceMethod.MethodName = methodEntry.Value.(string)
+						if resourceMethod.MethodName == "withId" {
+							resourceMethod.MethodName = "withID"
+						}
+					}
+
+					if methodKey == "get" || methodKey == "post" || methodKey == "delete" {
+						resourceHTTPMethod := ResourceHTTPMethod{
+							HTTPMethod: methodKey,
+						}
+						if methodEntry.Value != nil {
+							methodInfo := methodEntry.Value.(yaml.MapSlice)
+							if methodInfo != nil {
+								resourceHTTPMethod.Description = getPropertyString(methodInfo, "description")
+								if methodKey == "delete" {
+									isSection := getPropertyValue(methodInfo, "is")
+									if isSection != nil {
+										for _, traitInterface := range isSection.([]interface{}) {
+											if traitName, ok := traitInterface.(string); ok {
+												if traitName == "dataErasure" {
+													resourceHTTPMethod.DeleteHasDataErasure = true
+													break
+												}
+											}
```

Comment:
this is nested quite deep :-)

Reply:
Agreed, I refactored it (and also refactored the traits part).